lich.rbw

```diff
@@ -5846,10 +5856,24 @@ def respond(first = "", *messages)
 		elsif $frontend == 'profanity'
 			str = str.gsub('&', '&amp;').gsub('<', '&lt;').gsub('>', '&gt;')
 		end
-		wait_while { XMLData.in_stream }
-		$_CLIENT_.puts(str)
+		# Double-checked locking to avoid interrupting a stream and crashing the client
+		str_sent = false
+		if $_CLIENT_
+			until str_sent
+				wait_while { XMLData.in_stream }
+				str_sent = $_CLIENT_.puts_if(str) { !XMLData.in_stream }
+			end
```

Comment:
Would this be any clearer as `pause 0.1 until $_CLIENT_.puts_if(str) { !XMLData.in_stream }` where 0.1 is whatever arbitrary value works well.

Reply:
The problem with this is that each call to puts_if acquires the lock, which is theoretically expensive.  Also wait_while is set up to be efficient about context switching (it deprioritizes the current thread while waiting). 

The pattern I'm using here is a pretty standard version of https://en.wikipedia.org/wiki/Double-checked_locking , I think.

Reply:
Yeah no you're right, I forgot the mutex  lock was occurring before the yield down in put_if.